Sources/Purchasing/Purchases/TransactionPoster.swift

```diff
@@ -208,32 +304,39 @@ private extension TransactionPoster {
                            result: Result<(info: CustomerInfo, product: StoreProduct?), BackendError>,
                            subscriberAttributes: SubscriberAttribute.Dictionary?,
                            completion: @escaping CustomerAPI.CustomerInfoResponseHandler) {
-        let customerInfoResult = result.map(\.info)
-
-        self.operationDispatcher.dispatchOnMainActor {
-            switch result {
-            case let .success((customerInfo, product)):
-                if Self.shouldFinish(
-                    transaction: transaction,
-                    for: product,
-                    customerInfo: customerInfo
-                ) {
-                    self.finishTransactionIfNeeded(transaction) {
-                        completion(customerInfoResult)
-                    }
-                } else {
-                    completion(customerInfoResult)
+        let complete: @Sendable () -> Void = {
+            self.operationDispatcher.dispatchOnMainActor {
+                completion(result.map(\.info))
+            }
+        }
 
+        switch result {
+        case let .success((customerInfo, product)):
+            self.cache.savePostedTransaction(transaction)
+
+            if Self.shouldFinish(
+                transaction: transaction,
+                for: product,
+                customerInfo: customerInfo
+            ) {
+                self.finishTransactionIfNeeded(transaction) {
```

Comment:
A subtle change here, suggested by @aboedo: we no longer call `finishTransactionIfNeeded` from the main actor, we only call the completion (see `complete`) above block _after_ we're done with everything.